src/buffer.rs

```diff
@@ -0,0 +1,134 @@
+/// Special error returned when client attempts to add to full buffer.
+#[derive(Debug)]
+pub struct BufferFullError<T>(T);
+
+/// A buffer manages a fixed amount of data.
+///
+/// The data are produced one at a time or consumed all at once.
+pub trait Buffer<T> {
+    /// Add the given element to the buffer if possible.
+    fn add(&mut self, element: T) -> Result<(), BufferFullError<T>>;
+
+    /// Copy and return all elements of the buffer.
+    ///
+    /// When this operation completes, the buffer state will reset so that the buffer
+    /// has no elements. In other words, the buffer elements are completely consumed
+    /// and ownership transferred.
+    ///
+    /// Note that the underlying memory is *not* zeroed, so sensitive information should
+    /// not be stored in this buffer.
+    fn consume(&mut self) -> Option<Vec<T>>;
```

Comment:
This is fine for now, but I wonder how multiple observers will be able to read buffer data. Will the incoming client action queue be separate from this object-unique buffer of actions to apply to said object?

Reply:
I think we'll need separate Action and Update queues. The former increases response throughput, the latter reduces lock contention on the state.

But this could be prematurely optimizing. We might just have to stress test joined vs. separate queues and measure what works best.